cpp/src/arrow/util/decimal_test.cc

```diff
@@ -590,13 +597,62 @@ TEST(TestDecimalFromRealFloat, LargeValues) {
   // Test the entire float range
   for (int32_t scale = -38; scale <= 38; ++scale) {
     float real = std::pow(10.0f, static_cast<float>(scale));
-    CheckDecimalFromRealIntegerString(real, 1, -scale, "1");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 1, -scale, "1");
+  }
+  for (int32_t scale = -37; scale <= 36; ++scale) {
+    float real = 123.f * std::pow(10.f, static_cast<float>(scale));
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 2, -scale - 1, "12");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 3, -scale, "123");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 4, -scale + 1, "1230");
+  }
+}
+
+// Same tests, repeated for Decimal256
+class TestDecimal256FromRealFloat : public ::testing::TestWithParam<FromFloatTestParam> {
+};
+
+TEST_P(TestDecimal256FromRealFloat, SuccessConversion) {
+  const auto param = GetParam();
+  CheckDecimalFromReal<Decimal256>(param.real, param.precision, param.scale,
+                                   param.expected);
+}
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(
+    TestDecimal256FromRealFloat, TestDecimal256FromRealFloat,
+    ::testing::Values(
```

Comment:
Since those are the same values as for Decimal128, factor them out in a global variable?